### DISCUSSION TITLE
Attributes for similar experience of xUnit, AutoFixture & Moq?

### DISCUSSION THREAD

**Discussion by danielwertheim** [bchavez/Bogus#532] [Q&A] [answered]
I'm not so experienced with the Bogus ecosystem, but looking into it as a potential replacement for AutoFixture used with Moq and xUnit.

Are there any "official" or community supported projects that helps with getting support for a similar setup that you can achieve with AutoFixture, Moq and xUnit by decorating your tests with attributes?

E.g.

```csharp
[Theory]
[AutoData]
public void SomeTest(
    SomeDto dto,
    MyService sut)
{
    // dto: auto created and populated with data
    // sut: auto created (no dependencies)
}
```

```csharp
[Theory]
[CustomAutoData(typeof(MyCustomization)]
public void SomeTest(
    SomeDto dto,
    MyService sut)
{
    // CustomAutoData: Custom attribute to allow for passing generalized customizations to affect e.g. how properties are populated
    // dto: auto created and populated with data, with customizations from MyCustomization
    // sut: auto created (no dependencies)
}
```

```csharp
[Theory]
[AutoMoqData]
public void SomeTest(
    SomeDto dto,
    [Frozen] ISomeDependency dependencyOfSut,
    MyService sut)
{
    // AutoMoqData: Custom attribute that hooks in Moq, to automatically create mocked implementations of types
    // dto: auto created and populated with data
    // dependencyOfSut: auto mocked using (Moq) to have mocked return values etc
    // sut: auto created, with mocked dependency "dependencyOfSut"
    // [Frozen] "fixed" reference to the mocked dependency to interact with in the test
}
```

**Comment by bchavez** [comment, marked as answer]
Hello. Thanks for the question.

There's nothing "official"; but the community has developed a small ecosystem around **Bogus**. Most of the "community extensions" I'm familiar with are documented in the README here:
* https://github.com/bchavez/Bogus?tab=readme-ov-file#amazing-community-extensions
  and there's probably more that can be found on NuGet.

If you're looking for something to populate a DTO automatically without explicitly specifying rules; there's a community project called "**AutoBogus**":

* Original Work and Author: https://github.com/nickdodd79/AutoBogus
   * The original author of **AutoBogus** hasn't responded to a few issues in a while, so the community made a fork linked below:
   * Also, **AutoBogus** had some package integrations for Moq, FakeItEasy, etc.
* The recent fork of **AutoBogus** that seems to be gaining quite a bit of traction is: https://github.com/soenneker/soenneker.utils.autobogus
* There's also "**RoboBogus**" but I'm not that familiar with it and I can't find Simon's source-code: https://www.nuget.org/packages?q=AutoBogus&prerel=true&sortby=relevance

So the options above solve the "how to generate a DTO without explicitly specifying rules" part; and my guess is the next part is integrating with `[AutoData]` or `[AutoMoqData]` attributes with **xUnit**. I don't know of any integration package with **xUnit** or **Nunit** that leverages **AutoBogus** with attributes in the way you describe.

I'm only speculating, but I'm guessing that it is probably not too difficult to wire up `[AutoData]` attribute to give you a populated DTO since the most difficult part would be the autodata/autorules generation part which is already handled by **AutoBogus** et. al. My guess is, you'd only need to figure out the "integration part" that bridges the gap between **AutoBogus** and **xUnit** attributes.
